File: cracking-coding-interview/1/4.cpp
#include <iostream>
#include <cstring>

using std::cout;
using std::endl;

void ReplaceSpace(char *str, int true_length) {
    const char *replace_to = "%20";
    const int length_of_replace_to = 3;

    int num_spaces = 0;

    for (int i = 0; i < true_length; ++i)
        if (str[i] == ' ')
            ++num_spaces;
    // Let `p_source` point to the end of the string.
    char *p_source = &str[true_length-1];
    char *p_target = p_source + num_spaces * (length_of_replace_to - 1);

    while (num_spaces != 0) {
        if (*p_source == ' ') {
            --num_spaces;
            p_target -= length_of_replace_to;
            --p_source;
            char *p_temp = p_target;
            for (int i = 0; i < length_of_replace_to; ++i)
                *(++p_temp) = replace_to[i];
        } else {
            *(p_target--) = *(p_source--);
        }
    }
}

int main(int argc, char *argv[]) {
    char a[] = "", b[] = "hello", c[] = "hello world  ", d[] = " hello world       ";
    int length_a = 0, length_b = 5, length_c = 11, length_d = 13;
    ReplaceSpace(a, length_a);
    ReplaceSpace(b, length_b);
    ReplaceSpace(c, length_c);
    ReplaceSpace(d, length_d);

    cout << a << endl
        << b << endl
        << c << endl
        << d << endl;
}

File: cracking-coding-interview/1/4.py
def replace_space(byte_array, true_length):
    """ Inplace replace space in `byte_array`. `byte_array` is supposed to be
    long enough to hold the result string. """
    replace_to = '%20'
    replace_to_ord_list = [ord(x) for x in replace_to]

    space_count = 0
    space_ord = ord(' ')
    for x in byte_array[:true_length]:
        if x == space_ord:
            space_count += 1

    source_index = true_length - 1
    target_index = source_index + space_count * (len(replace_to) - 1)

    while space_count != 0:
        if byte_array[source_index] == space_ord:
            space_count -= 1
            byte_array[target_index - len(replace_to) + 1:target_index+1] = replace_to_ord_list
            source_index -= 1
            target_index -= len(replace_to)
        else:
            byte_array[target_index] = byte_array[source_index]
            source_index -= 1
            target_index -= 1

if __name__ == '__main__':
    a = bytearray(b''); length_a = 0
    b = bytearray(b'hello'); length_b = 5
    c = bytearray(b'hello world  '); length_c = 11
    d = bytearray(b' hello world       '); length_d = 13

    replace_space(a, length_a)
    replace_space(b, length_b)
    replace_space(c, length_c)
    replace_space(d, length_d)

    print(a, b, c, d, sep='\n')
